src/java/picard/sam/util/ReadNameParser.java

```diff
@@ -0,0 +1,177 @@
+                    this.warnedAboutRegexNotMatching = true;
+                }
+                return false;
+            }
+            loc.setTile((short) tmpLocationFields[0]);
+            loc.setX(tmpLocationFields[1]);
+            loc.setY(tmpLocationFields[2]);
+            return true;
+        } else if (this.readNameRegex == null) {
+            return false;
+        } else {
+            // Standard version that will use the regex
+            if (this.readNamePattern == null) this.readNamePattern = Pattern.compile(this.readNameRegex);
+
+            final Matcher m = this.readNamePattern.matcher(readName);
+            if (m.matches()) {
+                loc.setTile((short) Integer.parseInt(m.group(1)));
+                loc.setX(Integer.parseInt(m.group(2)));
+                loc.setY(Integer.parseInt(m.group(3)));
+                return true;
+            } else {
+                if (null != log && !this.warnedAboutRegexNotMatching) {
+                    this.log.warn(String.format("READ_NAME_REGEX '%s' did not match read name '%s'.  Your regex may not be correct.  " +
+                                    "Note that this message will not be emitted again even if other read names do not match the regex.",
+                            this.readNameRegex, readName));
+                    warnedAboutRegexNotMatching = true;
+                }
+                return false;
+            }
+        }
+    }
+
+    /**
+     * Given a string, splits the string by the delimiter, and returns the the last three fields parsed as integers.  Parsing a field
+     * considers only a sequence of digits up until the first non-digit character.  The three values are stored in the passed-in array.
+     */
+    public static int getLastThreeFields(final String readName, final char delim, final int[] tokens) {
+        int tokensIdx = 2; // start at the last token
+        int numFields = 0;
+        int i, endIdx;
+        endIdx = readName.length();
+        // find the last three tokens only
+        for (i = readName.length() - 1; 0 <= i && 0 <= tokensIdx; i--) {
+            if (readName.charAt(i) == delim || 0 == i) {
+                numFields++;
+                tokens[tokensIdx] = rapidParseInt(readName.substring((0 == i) ? 0 : (i+1), endIdx));
+                tokensIdx--;
+                endIdx = i;
+            }
+        }
+        // continue to find the # of fields
+        while (0 <= i) {
+            if (readName.charAt(i) == delim || 0 == i) numFields++;
+            i--;
+        }
+        if (numFields < 3) {
+            tokens[0] = tokens[1] = tokens[2] = -1;
+            return -1;
+        }
+        else {
+            return numFields;
+        }
+    }
+
+    /**
+     * Very specialized method to rapidly parse a sequence of digits from a String up until the first
+     * non-digit character.
+     */
+    public static int rapidParseInt(final String input) {
```

Comment:
Can I make one more suggestion here?  I think this method should fail if it cannot parse at least one digit from the string.  Looking at how this failed, it will in some instances have been parsing flowcell names and other non-numeric Strings, and happily returning 0.